### PR TITLE
`RealmDictionary` - 3: add support for `entries`

### PR DESCRIPTION
**Pull request by edualonso**
The `MutableMap.entries` function returns a `MutableSet<MutableMap.MutableEntry<K, V>>`. Said set is connected to the `RealmDictionary` from which it came and modifying it will result in those modifications being applied to the dictionary as well. Working with the dictionary's elements can be done via index access, which is supported by the C-API and Core.

Interestingly enough, the default implementations offered by JVM and K/N does support deletions on the mutable set, though they do not support additions since the function enabling this functionality isn't implemented and throws an `UnsupportedOperationException`. However, we can easily implement accessing the dictionary with a positioning index.

In order to add entries to the mutable set users need to create `MutableMap.MutableEntry` instances. A series of `realmMapEntryOf` helper functions have been added to allow this functionality.

It is worth noting that entries can be either managed or unmanaged, following the same pattern we have elsewhere.

Additional changes to the compiler plugin to fix an error that allowed `RealmDictionary<RealmObject>` fields in compile time which is incorrect since dictionaries of objects/embedded objects can only be nullable.

Pull request plan:
- [x] basic infrastructure and API: `RealmMap` and `RealmDictionary` interfaces and associated helper functions (stubs)
- [x] setting up exhaustive testing infrastructure, add support for `copyToRealm` for objects with dictionaries and interface functions `get`, `put`, `putAll`, `remove` and `clear` for all types except `RealmAny` (it will be done once the interface implementation is complete)
- [x] **THIS PR:** support for `entries` (i.e. `RealmMapEntrySet`, `UnmanagedRealmMapEntry` and `ManagedRealmMapEntry`) and internal entry set iterator (which enables iterating through the dictionary's k-v pairs externally)
- [ ] `containsKey`, `containsValue` - currently missing support in the C-API (https://github.com/realm/realm-core/issues/6181)
- [ ] `values` - this change entails refactoring `RealmResults` to contain primitive values
- [ ] support for `RealmAny`
- [ ] support for ~~`Deleteable` and~~ (dictionaries don't own objects like sets and lists do so removing them from the dictionary won't remove them from the realm) `asFlow` - currently blocked by https://github.com/realm/realm-core/issues/6228
- [ ] allow `ByteArray`s to be compared structurally to allow `remove` operations on entry set and values
- [ ] support for `copyFromRealm`
- [ ] support for dynamic realms
- [ ] sync testing
- [ ] documentation
- [ ] changelog entry

### CODE REVIEW

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt [hunk 1]
```diff
@@ -80,6 +81,8 @@ internal abstract class ManagedRealmMap<K, V>(
  */
 internal interface MapOperator<K, V> : CollectionOperator<V, RealmMapPointer> {
 
+    // Modification counter used to detect concurrent writes from the iterator
+    var modCount: Int
```
Comment (cmelchior) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt:85]: I suspect we need to make this field volatile to avoid threading issues on JVM... not sure if that is possible in common code, alternatively make it atomic? 

How is Kotlin solving this in the std library?

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt [hunk 2]
```diff
@@ -318,30 +327,202 @@ internal class ManagedRealmDictionary<E>(
...
+/**
+ * This class implements the typealias [RealmMapEntrySet] which matches
+ * `MutableSet<MutableMap.MutableEntry<K, V>>`. This class allows operating on a [ManagedRealmMap]
+ * in the form of a [Set] of [MutableMap.MutableEntry] values.
+ *
+ * Deletions are supported by the default semantics in JVM and K/N. These two operations are
+ * equivalent:
+ * ```
+ * dictionary.remove(myKey)
+ * dictionary.entries.remove(myKey, myValue) // implies we know the value of myValue
```
Comment (cmelchior) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt:339]: Does `entries.remove` really require both key and value? 🤯

File: packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/RealmDictionaryTests.kt [hunk 3]
```diff
@@ -489,6 +544,16 @@ internal interface DictionaryApiTester<T, Container> : ErrorCatcher {
...
+    fun entries_size()
+    fun entries_add()
+    fun entries_addAll()
+    fun entries_clear()
+    fun entries_iteratorNext() // This tests also hasNext
+    fun entries_iteratorNext_managedEntry_setValue()
+    fun entries_iteratorRemove()
+    fun entries_iteratorConcurrentModification()
+    fun entries_remove()
+    fun entries_removeAll()
```
Comment (cmelchior) [inline packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/RealmDictionaryTests.kt:556]: It it s a bit hard to follow the logic in Github, but is `equals/hashcode` in the `MapEntry` implementations being tested anywhere in this?

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt [hunk 4]
```diff
@@ -264,7 +273,7 @@ internal class RealmObjectMapOperator<K, V>(
                 )
                 Pair(previousObject, it.second)
             } as Pair<V?, Boolean>
-        }
+        }.also { modCount += 1 }
```
Comment (clementetb) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt:276]: Why not `modCount++`?

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt [hunk 5]
```diff
@@ -318,30 +327,207 @@ internal class ManagedRealmDictionary<E>(
...
+internal class ManagedRealmMapEntry<K, V>(
+    override val key: K,
+    private val operator: MapOperator<K, V>
+) : MutableMap.MutableEntry<K, V?> {
+
+    override val value: V?
+        get() = operator.get(key)
+
+    override fun setValue(newValue: V?): V? {
+        val previousValue = operator.get(key)
```
Comment (clementetb) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt:502]: We could skip the auxiliary variable by using `also`
```
return operator.get(key).also{
    operator.insert(key, newValue)
}
```

**Review by clementetb — COMMENTED** [review]
Some minor comments

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt (hunk 4 above)
Comment (edualonso) [reply packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt:276]: The `++` operator doesn't modify the instance on which it is executed, it returns the modified value 🤯 see https://kotlinlang.org/docs/operator-overloading.html#increments-and-decrements

File: packages/plugin-compiler/src/main/kotlin/io/realm/kotlin/compiler/AccessorModifierIrGeneration.kt [hunk 6]
```diff
@@ -1017,14 +1017,27 @@ class AccessorModifierIrGeneration(private val pluginContext: IrPluginContext) {
...
 
             // Lists of objects/embedded objects and sets of object may NOT contain null values, but dictionaries may
-            if (collectionType == CollectionType.SET || collectionType == CollectionType.LIST) {
-                if (isNullable) {
-                    logError(
-                        "Error in field ${declaration.name} - ${collectionType.description} does not support nullable realm objects element types.",
-                        declaration.locationOf()
-                    )
-                    return null
+            when (collectionType) {
```
Comment (edualonso) [inline packages/plugin-compiler/src/main/kotlin/io/realm/kotlin/compiler/AccessorModifierIrGeneration.kt:1020]: I had to make changes to this to fix an error that allowed `RealmDictionary<RealmObject>` fields in compile time which is incorrect since dictionaries of objects/embedded objects can only be nullable. Check also changes related to `ListTests`, `SetTests` and `DictionaryTests` as well.

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/ext/RealmDictionaryExt.kt [hunk 7]
```diff
@@ -41,31 +42,31 @@ public fun <T> realmDictionaryOf(elements: Collection<Pair<String, T>>): RealmDi
...
+    realmMapEntryOf(pair)
 
 /**
- * Instantiates an **unmanaged** [RealmMapMutableEntry] from a [key]-[value] pair. Entries are used
- * by the entry set produced by [RealmDictionary.entries]. It is possible to update the entry
- * which will result in the underlying [RealmDictionary] to be updated too.
+ * Instantiates an **unmanaged** [RealmMapMutableEntry] from a [key]-[value] pair
+ * that can be added to an entry set produced by [RealmDictionary.entries]. It is possible to add an
+ * unmanaged entry to a dictionary entry set. This will result in the underlying
+ * [RealmDictionary] being updated.
```
Comment (cmelchior) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/ext/RealmDictionaryExt.kt:57]:
```suggestion
 * unmanaged entry to a dictionary entry set. This will result in the entry being copied to Realm, 
 * updating the underlying [RealmDictionary].
```

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/ext/RealmDictionaryExt.kt [hunk 8]
```diff
@@ -41,31 +42,31 @@ public fun <T> realmDictionaryOf(elements: Collection<Pair<String, T>>): RealmDi
     }
 
 /**
- * Instantiates an **unmanaged** [RealmMapMutableEntry] from a [Pair] of [K] and [V]. Entries are
- * used by the entry set produced by [RealmDictionary.entries]. It is possible to update the entry
- * which will result in the underlying [RealmDictionary] to be updated too.
+ * Instantiates an **unmanaged** [RealmDictionaryMutableEntry] from a [Pair] of [String] and [V]
+ * that can be added to an entry set produced by [RealmDictionary.entries]. It is possible to add an
+ * unmanaged entry to a dictionary entry set. This will result in the underlying
+ * [RealmDictionary] being updated.
```
Comment (cmelchior) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/ext/RealmDictionaryExt.kt:48]:
```suggestion
 * unmanaged entry to a dictionary entry set. This will result in the entry being copied to Realm, 
 * updating the underlying [RealmDictionary].
```

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt [hunk 9]
```diff
@@ -318,30 +329,207 @@ internal class ManagedRealmDictionary<E>(
...
+/**
+ * This class implements the typealias [RealmMapEntrySet] which matches
+ * `MutableSet<MutableMap.MutableEntry<K, V>>`. This class allows operating on a [ManagedRealmMap]
+ * in the form of a [Set] of [MutableMap.MutableEntry] values.
+ *
+ * Deletions are supported by the default semantics in JVM and K/N. These two operations are
+ * equivalent:
+ * ```
+ * dictionary.remove(myKey)
+ * dictionary.entries.remove(myKey, myValue) // implies we know the value of myValue
```
Comment (cmelchior) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmMapInternal.kt:341]:
```suggestion
 * dictionary.entries.remove(realmDictionaryEntryOf(myKey, myValue)) // implies we know the value of myValue
```

File: packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/RealmDictionaryTests.kt [hunk 10]
```diff
@@ -606,6 +690,429 @@ internal abstract class ManagedDictionaryTester<T>(
...
+                // TODO revisit exception assertion once unified error handling is merged
+                assertFailsWith<RealmException> {
+                    entries.removeAll(shouldThrow)
+                }
+            }
+        }
+    }
+
+    override fun entry_equals() {
+        // Ignore ByteArray and RealmObject: structural equality cannot be assessed for these types
```
Comment (cmelchior) [inline packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/RealmDictionaryTests.kt:1081]: It doesn't look like these are being tested elsewhere? But I assume it is coming in a later PR.

**Review by cmelchior — APPROVED** [review]

File: packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/RealmDictionaryTests.kt (hunk 10 above)
Comment (edualonso) [reply packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/RealmDictionaryTests.kt:1081]: Correct. Equality for byte arrays is fixed in #1254

**Comment by edualonso** [comment]
I will merge it after updating the docs since the last build was successful.
